src/lib/awsConnectionParameters.ts

```diff
@@ -282,26 +282,78 @@ export async function getCredentials(awsParams: AWSConnectionParameters): Promis
 async function queryRegionFromMetadata(): Promise<string> {
     // SDK doesn't support discovery of region from EC2 instance metadata, so do it manually. We set
     // a timeout so that if the build host isn't EC2, we don't hang forever
-    return new Promise<string>((resolve, reject) => {
-        const metadataService = new AWS.MetadataService()
-        metadataService.httpOptions.timeout = 5000
-        metadataService.request('/latest/dynamic/instance-identity/document', (err, data) => {
-            try {
+    const imdsOptions = { httpOptions: { timeout: 5000 }, maxRetries: 2 }
+
+    const metadataService = new AWS.MetadataService(imdsOptions)
+
+    let token
+
+    try {
+        token = await getImdsV2Token(metadataService)
+    } catch {
+        // set token to empty -- this will use IMDSv1 mechanism to fetch region
+        token = ''
+    }
+
+    return (await getRegionFromImds(metadataService, token as string)) as string
+}
+
+/**
+ * Attempts to get a Instance Metadata Service V2 token
+ */
+async function getImdsV2Token(metadataService: AWS.MetadataService) {
+    console.log('Attempting to retrieve an IMDSv2 token.')
+
+    return new Promise((resolve, reject) => {
+        metadataService.request(
+            '/latest/api/token',
+            {
+                method: 'PUT',
+                headers: { 'x-aws-ec2-metadata-token-ttl-seconds': '21600' }
+            },
+            (err, token) => {
                 if (err) {
-                    throw err
+                    reject(err)
+                } else if (!token) {
+                    reject(new Error('IMDS did not return a token.'))
+                } else {
+                    resolve(token)
                 }
+            }
+        )
+    })
+}
+
+/**
+ * Attempts to get the region from the Instance Metadata Service
+ */
+async function getRegionFromImds(metadataService: AWS.MetadataService, token: string) {
```

Comment:
This eliminates the trailing `as string` on line 298

```suggestion
async function getRegionFromImds(metadataService: AWS.MetadataService, token: string): Promise<string> {
```